warehouse/query-core/src/main/java/datawave/query/jexl/visitors/AllTermsIndexedVisitor.java

```diff
@@ -63,111 +73,56 @@ public AllTermsIndexedVisitor(ShardQueryConfiguration config, MetadataHelper hel
-                    newIndex++;
-                }
-            }
-        }
-        
-        switch (newNode.jjtGetNumChildren()) {
-            case 0:
-                return null;
-            case 1:
-                JexlNode child = newNode.jjtGetChild(0);
-                JexlNodes.promote(newNode, child);
-                return child;
-            default:
-                return newNode;
-        }
+        return visitJunctionNode(node, data);
     }
     
     @Override
     public Object visit(ASTAndNode node, Object data) {
-        ASTAndNode newNode = new ASTAndNode(ParserTreeConstants.JJTANDNODE);
-        newNode.image = node.image;
+        return visitJunctionNode(node, data);
+    }
+    
+    private Object visitJunctionNode(JexlNode node, Object data) {
+        JexlNode copy = getCopyWithVisitedChildren(node, data);
         
-        int newIndex = 0;
-        for (int i = 0; i < node.jjtGetNumChildren(); i++) {
-            Node newChild = (Node) node.jjtGetChild(i).jjtAccept(this, data);
-            
-            if (newChild != null) {
-                // When we have an AND or OR
-                if ((newChild instanceof ASTOrNode || newChild instanceof ASTAndNode)) {
-                    // Only add that node if it actually has children
-                    if (0 < newChild.jjtGetNumChildren()) {
-                        newNode.jjtAddChild(newChild, newIndex);
-                        newIndex++;
-                    }
-                } else {
-                    // Otherwise, we want to add the child regardless
-                    newNode.jjtAddChild(newChild, newIndex);
-                    newIndex++;
-                }
-            }
-        }
-        switch (newNode.jjtGetNumChildren()) {
+        switch (copy.jjtGetNumChildren()) {
             case 0:
                 return null;
             case 1:
-                JexlNode child = newNode.jjtGetChild(0);
-                JexlNodes.promote(newNode, child);
+                JexlNode child = copy.jjtGetChild(0);
+                promote(copy, child);
                 return child;
             default:
-                return newNode;
+                return copy;
         }
     }
     
+    private JexlNode getCopyWithVisitedChildren(JexlNode node, Object data) {
+        // @formatter:off
+        JexlNode[] children = Arrays.stream(JexlNodes.children(node))
+                        .map(n -> (JexlNode) n.jjtAccept(this, data)) // Visit the node.
+                        .filter(Objects::nonNull) // Remove null nodes.
+                        .filter(n -> (!(n instanceof ASTAndNode) && !(n instanceof ASTOrNode)) || JexlNodes.isNotChildless(n)) // Remove empty junction nodes.
+                        .toArray(JexlNode[]::new);
+        // @formatter:on
+        JexlNode copy = JexlNodes.newInstanceOfType(node);
+        copy.image = node.image;
+        JexlNodes.children(copy, children);
+        return copy;
+    }
```

Comment:
This method is essentially the same as the copy method in the base RebuildingVisitor.  The exception to this is the filtering step where it drops AND or OR node without any children.  That *shouldn't* happen, and if it is happening that points to problems elsewhere in the code base.  So, if we assume that we won't encounter an AND or OR node without any children (which I think is a fairly safe bet) then we could just use the RebuildingVisitor's copy method.  Even if we *DO* encounter an AND or OR node with 0 children, the switch statement in the and/or node visitors will handle that by returning null, which gets dropped by the rebuilding visitor's copy method.  

Sorry for the wall of text... I think that using the visit methods of the parent RebuildingVisitor is a little bit cleaner and simplifies the code a bit.  So, in each place where you call 'getCopyWithVisitedChildren' you can instead call super.visit(node, data) which calls RebuildingVisitor.copy, which visits the entire subtree and creates copies.

Reply:
Fixed.